src/state_writer.rs

```diff
@@ -41,20 +41,19 @@ pub struct StateWriter {
 }
 
 impl StateCache {
-    fn new(val: Vec<u8>) -> Self {
+    fn new(val: &[u8]) -> Self {
         Self {
             init: None,
-            value: val,
+            value: val.to_vec(),
             dirty: false,
             deleted: false,
         }
     }
 
-    fn new_existing(val: Vec<u8>) -> Self {
-        let init = val.clone();
+    fn new_existing(val: &[u8]) -> Self {
         Self {
-            init: Some(init),
-            value: val,
+            init: Some(val.to_vec()),
+            value: val.to_vec(),
```

Comment:
Better to define `let val = val.to_vec()` outside of struct to not need to call `to_vec()` two times.

Reply:
I think, it is not possible without extra `to_vec` or `clone`, because both `init` and `value` need to move `val`.